atd-vzd/migrations/default/1709836840391_data_model_init/up.sql

```diff
@@ -0,0 +1,293 @@
+    crash_id integer not null references db.crashes (crash_id),
+    unit_type_id integer references db.unit_types (id)
+);
+
+create index on db.units (crash_id);
+
+---------------------------------------------
+--- Create crash triggers -------------------
+---------------------------------------------
+create or replace function db.cris_crash_insert_rows()
+returns trigger
+language plpgsql
+as
+$$
+BEGIN
+    RAISE NOTICE 'Inserting vz_crashes and crashes rows';
+    -- insert new (editable) vz record (only crash ID)
+    INSERT INTO db.vz_crashes (crash_id) values (new.crash_id);
+    -- insert new combined / official record
+    INSERT INTO db.crashes (crash_id, primary_address, road_type_id, latitude, longitude) values (
+        new.crash_id, new.primary_address, new.road_type_id, new.latitude, new.longitude
+    );
+    RETURN NULL;
+END;
+$$;
+
+create trigger insert_new_cris_crash_into_vz_crash
+after insert on db.cris_crashes
+for each row
+execute procedure db.cris_crash_insert_rows();
+
+create or replace function db.vz_crash_update()
+returns trigger
+language plpgsql
+as $$
+BEGIN
+    RAISE NOTICE 'Refreshing crash ID % due to vz_crashes update', new.crash_id;
+    UPDATE
+        db.crashes
+    SET
+        primary_address = COALESCE(new.primary_address, cris_crashes.primary_address),
+        road_type_id = COALESCE(new.road_type_id, cris_crashes.road_type_id),
+        latitude = COALESCE(new.latitude, cris_crashes.latitude),
+        longitude = COALESCE(new.longitude, cris_crashes.longitude)
+    FROM (
+        SELECT
+            *
+        FROM
+            db.cris_crashes where cris_crashes.crash_id = new.crash_id) AS cris_crashes
+WHERE
+    crashes.crash_id = new.crash_id;
+    RETURN NULL;
+END;
+$$;
+
+create trigger update_crash_from_vz_crash_update
+after update on db.vz_crashes
+for each row
+execute procedure db.vz_crash_update();
+
+create or replace function db.cris_crash_update()
+returns trigger
+language plpgsql
+as $$
+DECLARE
+   vz_record  record;
+BEGIN
+    RAISE NOTICE 'Updating crash ID % due to cris_crashes update', new.crash_id;
+
+    SELECT INTO vz_record *
+        FROM db.vz_crashes where crash_id = new.crash_id;
+
+    UPDATE
+        db.crashes
+    SET
+        primary_address = COALESCE(vz_record.primary_address, NEW.primary_address),
+        road_type_id = COALESCE(vz_record.road_type_id, NEW.road_type_id),
+        latitude = COALESCE(vz_record.latitude, NEW.latitude),
+        longitude = COALESCE(vz_record.longitude, NEW.longitude)
+    WHERE
+        crashes.crash_id = new.crash_id;
+    RETURN NULL;
+END;
+$$;
+
+create trigger update_crash_from_cris_update
+after update on db.cris_crashes
+for each row
+execute procedure db.cris_crash_update();
+
+
+
+create or replace function db.update_crash_location()
+returns trigger
+language plpgsql
+as $$
+begin
+    if (new.latitude is distinct from old.latitude OR new.longitude is distinct from old.longitude) THEN
```

Comment:
when it comes to whether a trigger actually needs to fire i like putting this logic in the creation of the trigger, that way the function doesnt even fire at all if not necessary :p

e.g. 
```
CREATE OR REPLACE TRIGGER update_crash_location_id_on_update
BEFORE UPDATE ON public.crashes
WHEN (OLD.latitude IS DISTINCT FROM NEW.latitude OR OLD.longitude IS DISTINCT FROM NEW.longitude) ...
```

Reply:
thank you! i thought the same thing as soon as i saw yours 🙏